core/src/main/java/com/linecorp/armeria/common/RequestContextAwareMinimalStage.java

```diff
@@ -0,0 +1,224 @@
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletionStage<Void> acceptEitherAsync(CompletionStage<? extends T> other, Consumer<? super T> action) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletionStage<Void> acceptEitherAsync(CompletionStage<? extends T> other, Consumer<? super T> action, Executor executor) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletionStage<Void> runAfterEither(CompletionStage<?> other, Runnable action) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletionStage<Void> runAfterEitherAsync(CompletionStage<?> other, Runnable action) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletionStage<Void> runAfterEitherAsync(CompletionStage<?> other, Runnable action, Executor executor) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public <U> CompletionStage<U> thenCompose(Function<? super T, ? extends CompletionStage<U>> fn) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public <U> CompletionStage<U> thenComposeAsync(Function<? super T, ? extends CompletionStage<U>> fn) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public <U> CompletionStage<U> thenComposeAsync(Function<? super T, ? extends CompletionStage<U>> fn, Executor executor) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public <U> CompletionStage<U> handle(BiFunction<? super T, Throwable, ? extends U> fn) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public <U> CompletionStage<U> handleAsync(BiFunction<? super T, Throwable, ? extends U> fn) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public <U> CompletionStage<U> handleAsync(BiFunction<? super T, Throwable, ? extends U> fn, Executor executor) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletionStage<T> whenComplete(BiConsumer<? super T, ? super Throwable> action) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletionStage<T> whenCompleteAsync(BiConsumer<? super T, ? super Throwable> action) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletionStage<T> whenCompleteAsync(BiConsumer<? super T, ? super Throwable> action, Executor executor) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletionStage<T> exceptionally(Function<Throwable, ? extends T> fn) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletableFuture<T> toCompletableFuture() {
+        return delegate;
```

Comment:
Calling `complete()` or `completeExceptionally()` on the returned future should not complete the minimal stage (`this`). How about:

```java
return delegate.thenApply(Function.identity());
```

.. so that `delegate` is not completed when a user does this:

```java
minimalStage.toCompletableFuture().complete(null);
```